pandas/core/frame.py

```diff
@@ -2634,6 +2634,52 @@ def transpose(self, *args, **kwargs):
 
     T = property(transpose)
 
+    # ----------------------------------------------------------------------
+    # Array Interface
+
+    # This is also set in IndexOpsMixin
+    # GH#23114 Ensure ndarray.__op__(DataFrame) returns NotImplemented
+    __array_priority__ = 1000
+
+    def __array__(self, dtype=None):
+        return com.values_from_object(self)
+
+    def __array_wrap__(self, result: np.ndarray,
+                       context: Optional[Tuple] = None) -> 'DataFrame':
+        """
+        We are called post ufunc; reconstruct the original object and dtypes.
+
+        Parameters
+        ----------
+        result : np.ndarray
+        context : tuple, optional
+
+        Returns
+        -------
+        DataFrame
+        """
+
+        d = self._construct_axes_dict(self._AXIS_ORDERS, copy=False)
+        result = self._constructor(result, **d)
+
+        # we try to cast extension array types back to the original
```

Comment:
I *really* don't think we should be trying to cast back here. There are just two many dtypes for which the `ufunc` isn't dtype preserving.

Rather, we should apply the ufunc to each block, which dispatches to each array. Then we infer the correct array / block dtype based on the output.

Reply:
it is necessary; the reason is that the ufuncs operate on the *dense* values, so we need to preserve the EA type (e.g. sparseness in this case); it doesn't actually hurt to do this and makes a nice experience.

yes doing this on a block-by-block basis is better, but this is quite tricky currently as the ufuncs are not applied that way, so until we fully support ``__array_ufunc__`` this is only / best way.

I think we *should* do a block-by-block array_ufunc, but it is much harder than it looks (mainly because we have lots of op overrides for various classes as a work-around now and these need to be changed *first*).